### PR TITLE
Fix episode runtime of episodes with the same file

### PR DESCRIPTION
**Pull request by phate89**
In episodes we can have a file connected to more than episode. 
Since we use the streamdetail to overvwrite runtime episodes if we have a multiepisode file all single episode right now have the full video length.
For example a file 
"Lost 1x01-02" file is 80 minutes long.
Episode 1 now have 80 minutes runtime and episode 2 80 minutes.
I want to replace it with the average length so the episode 1 will have 40 min and episode 2 40 min.

### CODE REVIEW

File: xbmc/video/VideoDatabase.cpp [hunk 1]
```diff
@@ -4905,11 +4911,22 @@ void CVideoDatabase::UpdateTables(int iVersion)
     m_pDS->exec("ALTER TABLE settings ADD VideoStream integer");
     m_pDS->exec("ALTER TABLE streamdetails ADD strVideoLanguage text");
   }
+
+  if (iVersion < 104)
+  {
+    m_pDS->query(PrepareSQL("SELECT GROUP_CONCAT(episode.idEpisode), episode.c%02d, COUNT(episode.idFile) FROM episode JOIN streamdetails ON streamdetails.idFile = episode.idFile AND streamdetails.iStreamType = 0 WHERE episode.c%02d = streamdetails.iVideoDuration GROUP BY episode.idFile HAVING COUNT(episode.idFile)>1", VIDEODB_ID_EPISODE_RUNTIME, VIDEODB_ID_EPISODE_RUNTIME));
```
Comment (Razzeee) [inline xbmc/video/VideoDatabase.cpp]: Could you break this up so it's a little bit easier to read? But not really a blocker.

**Comment by Razzeee** [comment]
Code seems fine, we should test this before we merge it.

**Comment by MilhouseVH** [comment]
MySQL seems happy enough with the `SELECT GROUP_CONCAT(episode.idEpisode)...` query (tested in MySQL Workbench).

**Comment by MilhouseVH** [comment]
Not sure if I'm seeing this right (or understanding this PR).

This is with VideoDB 104, ie. this PR:

s1

Episodes `Genesis (1)` and `Genesis (2)` are a single file, `S01E01E02`.

Total runtime for this file is 5563 seconds (92 mins).

The `GROUP_CONCAT()` query returns:
```
14454,14455 | 5563	 | 2
```
So shouldn't `Genesis (1)` and `Genesis(2)` now be showing 46 minutes (5563 / 2)?

**Comment by MilhouseVH** [comment]
OK, I can see the `c09` column in the `episode` table has updated correctly - the two episodes now have a runtime of 2781, however the GUI doesn't seem to reflect this...it still thinks each episode duration is 5563.

Also, when querying with JSON, both `runtime` and `streamdetails > video > duration` continue to return 5563 for both episode 1 and episode 2.

**Comment by MilhouseVH** [comment]
Hah. If JSON is queried for `runtime` alone, it returns 2781  for the each episode 1 & 2:

JSON Query:
```
{"jsonrpc": "2.0", "params": {"sort": {"order": "ascending", "method": "label"}, "season": 1, "tvshowid": 140, "properties": ["art", "file", "runtime"], "limits": {"start": 0, "end": 400}}, "method": "VideoLibrary.GetEpisodes", "id": "libEpisodes"}
```

JSON Result:
```
        "episodes": [
          {
            "art": {
              "season.poster": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/season01-poster.jpg/",
              "thumb": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/Season 1/Quantum Leap S01E01E02-thumb.jpg/",
              "tvshow.banner": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/banner.jpg/",
              "tvshow.clearart": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/clearart.png/",
              "tvshow.clearlogo": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/logo.png/",
              "tvshow.fanart": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/fanart.jpg/",
              "tvshow.landscape": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/landscape.jpg/",
              "tvshow.poster": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/poster.jpg/"
            },
            "episodeid": 14454,
            "file": "nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/Season 1/Quantum Leap S01E01E02.mkv",
            "label": "1x01. Genesis (1)",
            "runtime": 2781
          },
```
However if JSON is queried for `runtime` **and** `streamdetails` (which returns `duration`) then both `runtime` and `duration` are 5563:

JSON Query:
```
{"jsonrpc": "2.0", "params": {"sort": {"order": "ascending", "method": "label"}, "season": 1, "tvshowid": 140, "properties": ["art", "file", "runtime", "streamdetails"], "limits": {"start": 0, "end": 400}}, "method": "VideoLibrary.GetEpisodes", "id": "libEpisodes"}
```

JSON Result:
```
        "episodes": [
          {
            "art": {
              "season.poster": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/season01-poster.jpg/",
              "thumb": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/Season 1/Quantum Leap S01E01E02-thumb.jpg/",
              "tvshow.banner": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/banner.jpg/",
              "tvshow.clearart": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/clearart.png/",
              "tvshow.clearlogo": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/logo.png/",
              "tvshow.fanart": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/fanart.jpg/",
              "tvshow.landscape": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/landscape.jpg/",
              "tvshow.poster": "image://nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/poster.jpg/"
            },
            "episodeid": 14454,
            "file": "nfs://192.168.0.3/mnt/share/media/Video-Private/TVShows/Quantum Leap/Season 1/Quantum Leap S01E01E02.mkv",
            "label": "1x01. Genesis (1)",
            "runtime": 5563,
            "streamdetails": {
              "audio": [
                {
                  "channels": 2,
                  "codec": "ac3",
                  "language": "eng"
                }
              ],
              "subtitle": [
                {
                  "language": "eng"
                }
              ],
              "video": [
                {
                  "aspect": 1.3329999446868896,
                  "codec": "h264",
                  "duration": 5563,
                  "height": 480,
                  "language": "",
                  "stereomode": "",
                  "width": 720
                }
              ]
            }
          },
          {
```

The GUI consistently shows 92 mins (5563) for each episode.

**Comment by phate89** [comment]
I tested it with a skin that shows runtime on the list and was OK, same with json.  I guess when we request details that field is overwritten again. So we have a field that have different value depending on where we requires it. This needs a big fix up.. did anyone know if there's a real use for episode runtime? Tvdb has only series runtime..

**Comment by Razzeee** [comment]
Well at least tmdb is considering adding episode runtime. Not sure when or if it will happen.

**Comment by thezoggy** [comment]
also may need to double check:
* writing data to nfo.. correct info written? 
* rounding issue/divide by 0

**Comment by Razzeee** [comment]
@phate89 
any plans to revisit this?

**Comment by Rechi** [comment]
@phate89 ping

**Comment by MartijnKaijser** [comment]
I will close this PR. If there's interest please open a new one (with required changes if needed).

**Pull request closed without merging**